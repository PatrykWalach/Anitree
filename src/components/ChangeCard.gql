#import './ChangeCardBanner'

query ChangeCardQuery($id: Int) {
  Media(id: $id) {
    id
    ...ChangeCardBanner_media
  }
}
