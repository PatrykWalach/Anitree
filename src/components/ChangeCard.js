import { ChangeCardBanner_media } from './ChangeCardBanner'

import gql from 'graphql-tag'

export const ChangeCardQuery = gql`
  query ChangeCardQuery($id: Int) {
    Media(id: $id) {
      id
      ...ChangeCardBanner_media
    }
  }
  ${ChangeCardBanner_media}
`
