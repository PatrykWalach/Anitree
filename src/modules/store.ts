import * as modules from '.'
import { Store, plugin } from 'vuex-composition-api'

export const store =
  process.env.NODE_ENV === 'production'
    ? undefined
    : new Store({
        plugins: [plugin(Object.values(modules))],
      })
