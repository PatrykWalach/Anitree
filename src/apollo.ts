import { ApolloLink, concat } from 'apollo-link'
import { InMemoryCache, NormalizedCacheObject } from 'apollo-cache-inmemory'
import { PersistedData, PersistentStorage } from 'apollo-cache-persist/types'

import { ApolloClient } from 'apollo-client'
import { HttpLink } from 'apollo-link-http'

import { Variables } from '@/graphql/schema/media'
import Vue from 'vue'
import VueApollo from 'vue-apollo'
import { persistCache } from 'apollo-cache-persist'

import { store } from '@/store'

const link = new HttpLink({
  uri: 'https://graphql.anilist.co',
})

const middle = new ApolloLink((operation, forward) => {
  const token = store.getState().settings.token

  if (token) {
    operation.setContext({
      headers: {
        Authorization: `Bearer ${token}`,
      },
    })
  }
  if (forward) {
    return forward(operation)
  }
  return null
})

const cache = new InMemoryCache({
  cacheRedirects: {
    Query: {
      Media: (_, args, { getCacheKey }) =>
        getCacheKey({ __typename: 'Media', id: (args as Variables).id }),
    },
  },
})

// if (store.getState()settings..cacheApollo) {
persistCache({
  cache,
  storage: window.localStorage as PersistentStorage<
    PersistedData<NormalizedCacheObject>
  >,
})
// }

export const apollo = new ApolloClient({
  cache,
  link: concat(middle, link),
})

Vue.use(VueApollo)

export const apolloProvider = new VueApollo({
  defaultClient: apollo,
})
