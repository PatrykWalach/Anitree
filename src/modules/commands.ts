import {
  DeleteCommand,
  DeleteCommandConstructor,
} from './commands/DeleteCommand'
import { Module, State } from 'vuex-composition-api'
import { SaveCommand, SaveCommandConstructor } from './commands/SaveCommand'
import { Form } from '@/types'
import { FuzzyDate } from '@/graphql/schema/media'
import { MediaList } from '@/graphql/schema/mediaListCollection'
import { useSettings } from './settings'
import { watch } from '@vue/composition-api'

export interface Command {
  saveState?(): void | Promise<void>
  undo?(): void | Promise<void>
  execute(): void | Promise<void>
}

export interface ListCommand extends Command {
  backup: any | null
  variables: any
  undo(): Promise<void>
  execute(): Promise<void>
  done: boolean
  _class: string
}

export const isDeleteCommand = (
  e: (SaveCommandConstructor | DeleteCommandConstructor) &
    Pick<ListCommand, '_class'>,
): e is DeleteCommandConstructor & Pick<DeleteCommand, '_class'> =>
  e._class === 'DeleteCommand'

export const isSaveCommand = (
  e: (SaveCommandConstructor | DeleteCommandConstructor) &
    Pick<ListCommand, '_class'>,
): e is SaveCommandConstructor & Pick<SaveCommand, '_class'> =>
  e._class === 'SaveCommand'

const stored = localStorage.getItem('CHANGES')

export const loadHistory = (stored: string | null) =>
  stored &&
  JSON.parse(stored, (key, value) => {
    if (value instanceof Object) {
      if (isSaveCommand(value)) {
        return new SaveCommand(value)
      }
      if (isDeleteCommand(value)) {
        return new DeleteCommand(value)
      }
    }
    return value
  })

export const executeNotDoneCommands = async (commands: ListCommand[]) => {
  for (const command of commands.filter(({ done }) => !done)) {
    try {
      await command.execute()
    } catch {
      break
    }
  }
}

export const useCommands = (settings: ReturnType<typeof useSettings>) =>
  new Module({
    name: 'commands',
    namespaced: true,
    setup({ state, mutation }) {
      const history: State<ListCommand[]> = state(
        // loadHistory(stored) ||
        [],
      )

      // executeNotDoneCommands(history.value)

      // watch(() => {
      //   if (settings.state.cacheChanges.value) {
      //     localStorage.setItem('CHANGES', JSON.stringify(history.value))
      //   } else {
      //     localStorage.removeItem('CHANGES')
      //   }
      // })

      const PUSH_COMMAND = mutation(
        'PUSH_COMMAND',
        { history },
        (state, command: ListCommand) => state.history.push(command),
      )

      const POP_COMMAND = mutation('POP_COMMAND', { history }, state =>
        state.history.pop(),
      )

      const undo = async () => {
        history.value[history.value.length - 1].undo().then(POP_COMMAND)
      }

      const add = async (command: ListCommand) => {
        PUSH_COMMAND(command)
        await command.execute()
        return command
      }

      return {
        actions: {
          add,
          undo,
        },
        mutations: {
          POP_COMMAND,
          PUSH_COMMAND,
        },
        state: {
          history,
        },
      }
    },
  })

export const mediaListToForm = (
  mediaListEntry: MediaList | null,
  advancedScoring: string[],
): Form => {
  if (mediaListEntry) {
    const advancedScores = advancedScoring
      .map(key => mediaListEntry.advancedScores[key])
      .map(value => value || 0)

    delete mediaListEntry.__typename
    delete mediaListEntry.startedAt.__typename
    delete mediaListEntry.completedAt.__typename
    delete mediaListEntry.private
    delete mediaListEntry.id
    delete mediaListEntry.mediaId

    return {
      ...mediaListEntry,
      advancedScores,
    }
  }

  const advancedScores = advancedScoring.map(() => 0)
  return {
    advancedScores,
    completedAt: { day: null, month: null, year: null },
    notes: '',
    progress: 0,
    progressVolumes: 0,
    repeat: 0,
    score: 0,
    startedAt: { day: null, month: null, year: null },
    status: null,
  }
}

export const isDate = (e: any): e is FuzzyDate => e.__typename === 'FuzzyDate'
