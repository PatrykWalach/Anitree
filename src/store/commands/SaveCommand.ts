import { ListCommand, mediaListToForm, useCommands } from '../commands'
import { MEDIA, SAVE_MEDIA_LIST_ENTRY, VIEWER, apollo } from '@/graphql'
import { Media, Variables as MediaVariables } from '@/graphql/schema/media'

import { Form } from '@/types'
import { SaveVariables } from '@/graphql/schema/listEntry'
import { User } from '@/graphql/schema/viewer'

export type SaveCommandConstructor = Partial<
  Pick<SaveCommand, 'backup' | 'done'>
> &
  Pick<SaveCommand, 'variables'>

export class SaveCommand implements ListCommand {
  public backup: Form | null
  public variables: SaveVariables
  public done: boolean
  public _class = 'SaveCommand' as const

  constructor({
    variables,

    backup = null,
    done = false,
  }: SaveCommandConstructor) {
    this.variables = variables
    this.backup = backup
    this.done = done
  }

  public saveState() {
    const { variables } = this

    return Promise.all([
      apollo
        .query<{ Media: Media }, MediaVariables>({
          query: MEDIA,
          variables: { id: variables.mediaId },
        })
        .then(({ data }) => data.Media),
      apollo
        .query<{ Viewer: User }>({ query: VIEWER })
        .then(({ data }) => data.Viewer),
    ] as [Promise<Media>, Promise<User>]).then(
      ([{ mediaListEntry, type }, { mediaListOptions }]) => {
        this.backup = Object.fromEntries(
          Object.entries(
            mediaListToForm(
              mediaListEntry,
              mediaListOptions[type === 'MANGA' ? 'mangaList' : 'animeList']
                .advancedScoring,
            ),
          ).filter(([key]) => variables.hasOwnProperty(key)),
        ) as Form
      },
    )
  }

  public async undo() {
    const {
      backup,
      variables: { mediaId },
    } = this

    if (this.done) {
      await apollo
        .mutate<MediaList, SaveVariables>({
          mutation: SAVE_MEDIA_LIST_ENTRY,
          variables: { mediaId, ...backup },
        })
        .then(() => {
          useCommands().history.value.pop()
        })
    } else {
      useCommands().history.value.pop()
    }
  }

  public async execute() {
    await this.saveState()

    const { variables } = this

    return apollo
      .mutate<MediaList, SaveVariables>({
        mutation: SAVE_MEDIA_LIST_ENTRY,
        variables,
      })
      .then(() => {
        this.done = true
      })
  }
}
