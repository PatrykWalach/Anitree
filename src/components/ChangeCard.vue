<template>
  <BaseQuery
    :apollo="{
      Media: {
        ...Media,
        tag: null,
      },
    }"
    v-slot="{ Media }"
  >
    <v-banner single-line>
      <span :title="getTitle(Media && Media.title)">
        {{ getTitle(Media && Media.title) }}
      </span>

      <template v-slot:icon>
        <v-tooltip top>
          <template v-slot:activator="{ on, attrs }">
            <v-icon
              :color="command.done ? 'primary' : 'error'"
              v-bind="attrs"
              v-on="on"
              >{{ command.done ? 'cloud_done' : 'error' }}</v-icon
            >
          </template>
          <span>{{ command.done ? 'Done' : 'Pending' }}</span>
        </v-tooltip>
      </template>
      <template v-slot:actions>
        <v-dialog max-width="440px">
          <template v-slot:activator="{ on, attrs }">
            <v-btn v-bind="attrs" text color="accent" v-on="on">
              More
            </v-btn>
          </template>
          <v-card>
            <v-simple-table>
              <thead>
                <tr>
                  <th>Changes</th>
                  <th>From</th>
                  <th>To</th>
                </tr>
              </thead>
              <tbody>
                <tr v-for="key in keys" :key="key">
                  <th class="text-capitalize">{{ key }}</th>
                  <td class="error--text">
                    <BaseTime
                      v-if="isDate(command.backup[key])"
                      :date="command.backup[key]"
                    />
                    <template v-else>
                      {{ command.backup[key] }}
                    </template>
                  </td>
                  <td class="success--text">
                    <BaseTime
                      v-if="isDate(command.variables[key])"
                      :date="command.variables[key]"
                    />
                    <template v-else>
                      {{ command.variables[key] }}
                    </template>
                  </td>
                </tr>
              </tbody>
            </v-simple-table>
          </v-card>
        </v-dialog>
      </template>
    </v-banner>
  </BaseQuery>
</template>
<script lang="ts">
import { computed, createComponent } from '@vue/composition-api'
import BaseQuery from './BaseQuery.vue'
import { DeleteCommand } from '@/modules/commands/DeleteCommand'
import { FuzzyDate } from '../graphql/schema/media'
import { SaveCommand } from '@/modules/commands/SaveCommand'
import { useMedia } from '@/graphql'

const BaseTime = () => import('./BaseTime.vue')

interface Props {
  command: SaveCommand | DeleteCommand
}

export default createComponent<Readonly<Props>>({
  components: {
    BaseQuery,
    BaseTime,
  },
  props: {
    command: {
      default: null,
      required: true,
      type: [SaveCommand, DeleteCommand],
    },
  },
  setup(props, { root }) {
    const {
      getters: { getTitle },
    } = root.$modules.title

    const isDate = (e: any): e is Omit<FuzzyDate, '__typename'> =>
      e instanceof Object &&
      e.hasOwnProperty('month') &&
      e.hasOwnProperty('day') &&
      e.hasOwnProperty('year')

    const keys = computed(() => {
      const {
        command: { variables: _variables, backup: _backup },
      } = props

      const variables = Object.assign({}, _variables)
      const backup = Object.assign({}, _backup)

      delete variables.id
      delete variables.mediaId

      if (!backup) {
        return Object.keys(variables).sort()
      }

      return [
        ...new Set(Object.keys(variables).concat(Object.keys(backup))),
      ].sort()
    })

    return {
      getTitle,
      isDate,
      keys,
      ...useMedia(() => ({ id: props.command.variables.mediaId })),
    }
  },
})
</script>
