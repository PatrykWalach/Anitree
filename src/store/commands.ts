import CompositionApi, { Ref, ref, watch } from '@vue/composition-api'
import {
  DeleteCommand,
  DeleteCommandConstructor,
} from './commands/DeleteCommand'
import { DeleteVariables, SaveVariables } from '@/graphql/schema/listEntry'
import { SaveCommand, SaveCommandConstructor } from './commands/SaveCommand'
import { Form } from '@/types'
import { FuzzyDate } from '@/graphql/schema/media'
import { MediaList } from '@/graphql/schema/mediaListCollection'
import Vue from 'vue'
import { useSettings } from './settings'

Vue.use(CompositionApi)

export interface Command {
  saveState?(): void | Promise<void>
  undo?(): void | Promise<void>
  execute(): void | Promise<void>
}

export interface ListCommand extends Command {
  backup: Form | null
  variables: SaveVariables | (DeleteVariables & Pick<SaveVariables, 'mediaId'>)
  done: boolean
  _class: 'SaveCommand' | 'DeleteCommand'
}

export const isDeleteCommand = (
  e: (SaveCommandConstructor | DeleteCommandConstructor) &
    Pick<ListCommand, '_class'>,
): e is DeleteCommandConstructor & Pick<DeleteCommand, '_class'> =>
  e._class === 'DeleteCommand'

export const isSaveCommand = (
  e: (SaveCommandConstructor | DeleteCommandConstructor) &
    Pick<ListCommand, '_class'>,
): e is SaveCommandConstructor & Pick<SaveCommand, '_class'> =>
  e._class === 'SaveCommand'

const stored = localStorage.getItem('CHANGES')

const history: Ref<ListCommand[]> = ref(
  (stored &&
    JSON.parse(stored, (key, value) => {
      if (value instanceof Object) {
        if (isSaveCommand(value)) {
          return new SaveCommand(value)
        }
        if (isDeleteCommand(value)) {
          return new DeleteCommand(value)
        }
      }
      return value
    })) ||
    [],
)

watch(() => {
  if (useSettings().cacheChanges.value) {
    localStorage.setItem('CHANGES', JSON.stringify(history.value))
  }
})
;(async () => {
  for (const command of history.value.filter(({ done }) => !done)) {
    await command.execute()
  }
})()

const dispatch = async (command: ListCommand) => {
  history.value.push(command)
  await command.execute()
  return command
}

export const useCommands = () => {
  return {
    dispatch,
    history,
  }
}

export const mediaListToForm = (
  mediaListEntry: MediaList | null,
  advancedScoring: string[],
): Form => {
  if (mediaListEntry) {
    const advancedScores = advancedScoring
      .map(key => mediaListEntry.advancedScores[key])
      .map(value => value || 0)

    delete mediaListEntry.__typename
    delete mediaListEntry.startedAt.__typename
    delete mediaListEntry.completedAt.__typename
    delete mediaListEntry.private
    delete mediaListEntry.id
    delete mediaListEntry.mediaId

    return {
      ...mediaListEntry,
      advancedScores,
    }
  }

  const advancedScores = advancedScoring.map(() => 0)
  return {
    advancedScores,
    completedAt: { day: null, month: null, year: null },
    notes: '',
    progress: 0,
    progressVolumes: 0,
    repeat: 0,
    score: 0,
    startedAt: { day: null, month: null, year: null },
    status: null,
  }
}

export const isDate = (e: any): e is FuzzyDate => e.__typename === 'FuzzyDate'
