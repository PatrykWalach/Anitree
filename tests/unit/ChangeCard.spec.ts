import ChangeCard from '@/components/ChangeCard.vue'

import { SaveCommand } from '@/store/modules/commands/SaveCommand'

import { matchSnapshot } from './utils'
import { mockedDollarApollo } from './mocks/apollo'
import { useMockedStore } from './mocks/store'

describe('ChangeCard', () => {
  matchSnapshot(ChangeCard, {
    propsData: {
      command: new SaveCommand({
        apollo: mockedDollarApollo,
        variables: {
          mediaId: 1,
        },
      }),
    },
    ...useMockedStore(),
  })
})
