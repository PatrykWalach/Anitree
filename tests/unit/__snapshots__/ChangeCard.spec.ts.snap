// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangeCard matches snapshot 1`] = `
<basequery-stub
  apollo="[object Object]"
/>
`;
